analyzers/src/SonarAnalyzer.CSharp/Rules/LoopsAndLinq.cs

```diff
@@ -43,15 +48,96 @@ public sealed class LoopsAndLinq : SonarDiagnosticAnalyzer
         protected override void Initialize(SonarAnalysisContext context) =>
             context.RegisterSyntaxNodeActionInNonGenerated(c =>
                 {
-                    var node = (ForEachStatementSyntax)c.Node;
-
-                    if (node.Statement is BlockSyntax blockSyntax
-                        && blockSyntax.ChildNodes().Count() == 1
-                        && blockSyntax.ChildNodes().Single() is IfStatementSyntax)
+                    var forEachStatementSyntax = (ForEachStatementSyntax)c.Node;
+                    if (CanBeSimplifiedUsingWhere(forEachStatementSyntax.Statement))
+                    {
+                        c.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, forEachStatementSyntax.Expression.GetLocation(), WhereMessageFormat));
+                    }
+                    else
                     {
-                        c.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, node.GetLocation()));
+                        CheckIfCanBeSimplifiedUsingSelect(forEachStatementSyntax, c);
                     }
                 },
                 SyntaxKind.ForEachStatement);
+
+        private static bool CanBeSimplifiedUsingWhere(SyntaxNode statement) =>
+            GetIfStatement(statement) is { } ifStatementSyntax
+            && CanIfStatementBeMoved(ifStatementSyntax);
+
+        private static IfStatementSyntax GetIfStatement(SyntaxNode node) =>
+            node switch
+            {
+                IfStatementSyntax ifStatementSyntax => ifStatementSyntax,
+                BlockSyntax blockSyntax when blockSyntax.ChildNodes().Count() == 1 => GetIfStatement(blockSyntax.ChildNodes().Single()),
+                _ => null
+            };
+
+        private static bool CanIfStatementBeMoved(IfStatementSyntax ifStatementSyntax) =>
+            ifStatementSyntax.Else == null
+            && !ContainsExitPoints(ifStatementSyntax)
+            && ifStatementSyntax.Condition is InvocationExpressionSyntax invocationExpressionSyntax
+            && !invocationExpressionSyntax.DescendantNodes().OfType<ArgumentSyntax>().Any(argument => argument.RefOrOutKeyword.IsKind(SyntaxKind.OutKeyword));
```

Comment:
what if the method is invoked with a `ref` argument? IMO it might be lost (harder to read) doing this inside a LINQ statement (although feel free to disagree, I don't have a strong opinion here)

Reply:
I've excluded `out` since it could lead to un-compilable code. With `ref` the code will compile so it's usually up for the developer to decide; a matter of preferences.

Now I think that it's better to exclude it to avoid annoying messages for the users who don't want to pass `ref` parameters to the invocations inside LINQ functions.